test/models/host_test.rb

```diff
@@ -3749,34 +3766,38 @@ def to_managed!
   describe '#uptime_seconds' do
     test 'should return uptime in seconds' do
       host = FactoryBot.create(:host)
-      fact = FactoryBot.create(:fact_name, name: 'ansible_uptime_seconds')
-      FactoryBot.create(:fact_value, fact_name: fact, host: host, :value => 123)
-      assert_equal 123, host.uptime_seconds
+      boot_time = 1.day.ago
+      host.create_reported_data(:boot_time => boot_time)
+      now = Time.zone.now.to_i
+      assert_equal host.uptime_seconds, now - boot_time.to_i
     end
 
     test 'should return nil if no uptime fact is available' do
       host = FactoryBot.create(:host)
       assert_nil host.uptime_seconds
     end
+  end
 
-    test 'should return uptime and based on backup facts even if there are multiple other facts' do
+  describe '#uptime_seconds' do
+    test 'should not fail on host without reported data' do
       host = FactoryBot.create(:host)
-      ansible_uptime_fact = FactoryBot.create(:fact_name, name: 'ansible_uptime_seconds', :type => 'FactName::Ansible')
-      chef_uptime_fact = FactoryBot.create(:fact_name, name: 'uptime_seconds', :type => 'FactName::Chef')
-      unrelated_fact = FactoryBot.create(:fact_name, name: 'os')
-      puppet_fact = FactoryBot.create(:fact_name, name: 'system_uptime::seconds')
-      ansible_fact = FactoryBot.create(:fact_value, fact_name: ansible_uptime_fact, host: host, :value => 123)
-      cf = FactoryBot.create(:fact_value, fact_name: chef_uptime_fact, host: host, :value => 222)
-      cf.update_attribute :updated_at, cf.updated_at + 1.second
-      FactoryBot.create(:fact_value, fact_name: unrelated_fact, host: host, :value => 'Fedora 29')
-      assert_equal 222, host.uptime_seconds
-      pf = FactoryBot.create(:fact_value, fact_name: puppet_fact, host: host, :value => 456)
-      pf.update_attribute :updated_at, cf.updated_at + 2.seconds
-      host.reload
-      assert_equal 456, host.uptime_seconds
-      ansible_fact.update :value => 789, :updated_at => cf.updated_at + 3.seconds
-      host.reload
-      assert_equal 789, host.uptime_seconds
+      assert_nothing_raised do
+        host.clear_data_on_build
+      end
+    end
+
+    test 'should delete reported data on rebuild' do
+      host = FactoryBot.create(:host)
+      boot_time = 1.day.ago
+      host.create_reported_data(:boot_time => boot_time)
+      refute_nil host.uptime_seconds
+      host.clear_data_on_build
```

Comment:
I believe, we need to reload after this, the record is destroyed, but still loaded.

Reply:
in fact it needed more tweaking so that clear_data_on_build would actually work (setting old and build)